okhttp/src/test/java/okhttp3/ConscryptTest.kt

```diff
@@ -15,34 +15,129 @@
  */
 package okhttp3
 
+import mockwebserver3.MockResponse
+import mockwebserver3.MockWebServer
 import okhttp3.TestUtil.assumeNetwork
 import okhttp3.internal.platform.ConscryptPlatform
 import okhttp3.internal.platform.Platform
 import okhttp3.testing.PlatformRule
+import okhttp3.tls.internal.TlsUtil
+import okio.ByteString.Companion.toByteString
 import org.assertj.core.api.Assertions.assertThat
 import org.conscrypt.Conscrypt
+import org.junit.jupiter.api.Assertions.assertEquals
 import org.junit.jupiter.api.Assertions.assertFalse
 import org.junit.jupiter.api.Assertions.assertTrue
 import org.junit.jupiter.api.BeforeEach
 import org.junit.jupiter.api.Disabled
 import org.junit.jupiter.api.Test
 import org.junit.jupiter.api.extension.RegisterExtension
+import org.junit.jupiter.params.ParameterizedTest
+import org.junit.jupiter.params.provider.ValueSource
+import javax.net.ssl.SSLSocket
 
 class ConscryptTest {
   @JvmField @RegisterExtension val platform = PlatformRule.conscrypt()
   @JvmField @RegisterExtension val clientTestRule = OkHttpClientTestRule()
 
-  private val client = clientTestRule.newClient()
+  private var client = clientTestRule.newClient()
 
-  @BeforeEach fun setUp() {
+  private val handshakeCertificates = TlsUtil.localhost()
+
+  private lateinit var server: MockWebServer
+
+  @BeforeEach @Throws(Exception::class) fun setUp(server: MockWebServer) {
     platform.assumeConscrypt()
+    this.server = server
+  }
+
+  private fun enableTls() {
+    client = client.newBuilder()
+      .sslSocketFactory(
+        handshakeCertificates.sslSocketFactory(), handshakeCertificates.trustManager
+      )
+      .build()
+    server.useHttps(handshakeCertificates.sslSocketFactory(), false)
   }
 
   @Test
   fun testTrustManager() {
     assertThat(Conscrypt.isConscrypt(Platform.get().platformTrustManager())).isTrue()
   }
 
+  @ParameterizedTest(name = "{displayName}({arguments})")
+  @ValueSource(strings = ["TLSv1.2", "TLSv1.3"])
+  fun testSessionReuse(tlsVersion: String) {
+    val sessionIds = mutableListOf<String>()
+
+    enableTls()
+
+    val tlsVersion = TlsVersion.forJavaName(tlsVersion)
+    val spec = ConnectionSpec.Builder(ConnectionSpec.MODERN_TLS)
+      .tlsVersions(tlsVersion)
+      .build()
+
+    var reuseSession = false
+
+    val sslContext = handshakeCertificates.sslContext()
+    val systemSslSocketFactory = sslContext.socketFactory
+    val sslSocketFactory = object : DelegatingSSLSocketFactory(systemSslSocketFactory) {
+      override fun configureSocket(sslSocket: SSLSocket): SSLSocket {
+        return sslSocket.apply {
+          if (reuseSession) {
```

Comment:
I find this somewhat hard to follow ... capturing a mutable local. Maybe I’m too used to how it works in Java and not Kotlin!

Reply:
I'd argue that it's the most natural way to express it in Kotlin test code.  Creating additional one off fake classes with these features seems like a step back. I've used things like an AtomicBoolean to allow a mutable capture of a local variable in Java.

And with things like inlined functions making it less clear when you look at the call site whether a DSL is actually all in the same scope or has function calls.